### PR TITLE
Move GetFirstBeforeTime/GetFirstAfterTime to TimerData

### PR DESCRIPTION
**Pull request by dimitry-**
And add unit-tests for them

Test: build, run ClientDataTests

### CODE REVIEW

File: src/ClientData/TimerDataTest.cpp [hunk 1]
```diff
@@ -66,4 +68,92 @@ TEST(TimerData, AddTimers) {
...
+
+  {
+    const TimerInfo* timer_info = timer_data.GetFirstBeforeTime(4, 0);
+    ASSERT_NE(timer_info, nullptr);
+    EXPECT_EQ(timer_info->start(), 2);
+    EXPECT_EQ(timer_info->end(), 5);
+  }
+
+  {
+    const TimerInfo* timer_info = timer_data.GetFirstBeforeTime(2, 0);
```
Comment (karupayun) [inline src/ClientData/TimerDataTest.cpp:142]: FirstBeforeTime include the time queried but FirstAfterTime not? Why?

File: src/ClientData/TimerDataTest.cpp [hunk 2]
```diff
@@ -66,4 +68,92 @@ TEST(TimerData, AddTimers) {
...
+
+  {
+    const TimerInfo* timer_info = timer_data.GetFirstBeforeTime(6, 0);
+    ASSERT_NE(timer_info, nullptr);
+    EXPECT_EQ(timer_info->start(), 2);
+    EXPECT_EQ(timer_info->end(), 5);
+  }
+
+  {
+    const TimerInfo* timer_info = timer_data.GetFirstBeforeTime(4, 0);
```
Comment (karupayun) [inline src/ClientData/TimerDataTest.cpp:135]: I would like to have an additional query for FirstBeforeTime that returns the last timer.

**Review by karupayun — APPROVED** [review]
LGTM!

**Comment by karupayun** [comment]
Only a comment. ThreadTracks has queries inside the ScopeTree, so they would work differently. Do we agree on that?

File: src/ClientData/include/ClientData/TimerData.h [hunk 3]
```diff
@@ -66,6 +66,44 @@ class TimerData final {
...
+    const orbit_client_data::TimerChain* chain = GetChain(depth);
+    if (chain == nullptr) return nullptr;
+
+    const orbit_client_protos::TimerInfo* first_timer_before_time = nullptr;
+
+    // TODO: do better than linear search...
+    for (const auto& it : *chain) {
+      for (size_t k = 0; k < it.size(); ++k) {
+        const orbit_client_protos::TimerInfo* timer_info = &it[k];
+        if (timer_info->start() > time) {
```
Comment (karupayun) [inline src/ClientData/include/ClientData/TimerData.h:97]: I suppose we should make >= here **or** in the other function, so both of them work in the same way?

File: src/ClientData/include/ClientData/TimerData.h [hunk 4]
```diff
@@ -66,6 +66,44 @@ class TimerData final {
 
   void UpdateMaxDepth(uint32_t depth) { max_depth_ = std::max(max_depth_, depth); }
 
+  [[nodiscard]] const orbit_client_protos::TimerInfo* GetFirstAfterTime(uint64_t time,
+                                                                        uint32_t depth) const {
+    const orbit_client_data::TimerChain* chain = GetChain(depth);
+    if (chain == nullptr) return nullptr;
+
+    // TODO: do better than linear search...
```
Comment (reichlfl) [inline src/ClientData/include/ClientData/TimerData.h:74]: Can you add a bug and link in the comment?

File: src/ClientData/include/ClientData/TimerData.h [hunk 5]
```diff
@@ -66,6 +66,44 @@ class TimerData final {
...
+  }
+
+  [[nodiscard]] const orbit_client_protos::TimerInfo* GetFirstBeforeTime(uint64_t time,
+                                                                         uint32_t depth) const {
+    const orbit_client_data::TimerChain* chain = GetChain(depth);
+    if (chain == nullptr) return nullptr;
+
+    const orbit_client_protos::TimerInfo* first_timer_before_time = nullptr;
+
+    // TODO: do better than linear search...
```
Comment (reichlfl) [inline src/ClientData/include/ClientData/TimerData.h:93]: Same here

File: src/ClientData/include/ClientData/TimerData.h [hunk 6]
```diff
@@ -66,6 +66,44 @@ class TimerData final {
 
   void UpdateMaxDepth(uint32_t depth) { max_depth_ = std::max(max_depth_, depth); }
 
+  [[nodiscard]] const orbit_client_protos::TimerInfo* GetFirstAfterTime(uint64_t time,
```
Comment (reichlfl) [inline src/ClientData/include/ClientData/TimerData.h:69]: Nit: Would this be an opportunity to rename the methods to `GetFirstAfterStartTime`? I find it confusing that we don't specify if we're using start- or end-times for such operations, but maybe that's just me.

**Review by reichlfl — APPROVED** [review]
LGTM!

File: src/ClientData/TimerDataTest.cpp (hunk 2 above)
Comment (dimitry-) [reply src/ClientData/TimerDataTest.cpp:135]: done. and yes - there was a bug.

File: src/ClientData/TimerDataTest.cpp (hunk 1 above)
Comment (dimitry-) [reply src/ClientData/TimerDataTest.cpp:142]: fixed.

File: src/ClientData/include/ClientData/TimerData.h (hunk 3 above)
Comment (dimitry-) [reply src/ClientData/include/ClientData/TimerData.h:97]: done.

File: src/ClientData/include/ClientData/TimerData.h (hunk 4 above)
Comment (dimitry-) [reply src/ClientData/include/ClientData/TimerData.h:74]: done.

File: src/ClientData/include/ClientData/TimerData.h (hunk 5 above)
Comment (dimitry-) [reply src/ClientData/include/ClientData/TimerData.h:93]: done

File: src/ClientData/include/ClientData/TimerData.h (hunk 6 above)
Comment (dimitry-) [reply src/ClientData/include/ClientData/TimerData.h:69]: done.

**Pull request merged**